File: app/src/main/java/com/gnardini/tvshowcontroller/directories/DirectoryAdapter.java
package com.gnardini.tvshowcontroller.directories;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.gnardini.tvshowcontroller.R;

import java.util.LinkedList;
import java.util.List;

public class DirectoryAdapter extends RecyclerView.Adapter<DirectoryAdapter.DirectoryViewHolder> {

    private DirectoryClickListener listener;
    private List<String> directories;

    public DirectoryAdapter(DirectoryClickListener listener) {
        this.listener = listener;
        this.directories = new LinkedList<>();
    }

    @Override
    public DirectoryViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater
                .from(parent.getContext())
                .inflate(R.layout.directory_view, parent, false);
        return new DirectoryViewHolder(view, listener);
    }

    @Override
    public void onBindViewHolder(DirectoryViewHolder holder, int position) {
        holder.setText(directories.get(position));
    }

    @Override
    public int getItemCount() {
        return directories.size();
    }

    public void setDirectories(List<String> directories) {
        this.directories = directories;
        notifyDataSetChanged();
    }

    static class DirectoryViewHolder extends RecyclerView.ViewHolder {

        private DirectoryClickListener listener;
        TextView directory;
        String text;

        public DirectoryViewHolder(View itemView, final DirectoryClickListener listener) {
            super(itemView);
            this.listener = listener;
            directory = itemView.findViewById(R.id.directory);
            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    listener.onDirectoryClicked(text);
                }
            });
        }

        public void setText(String text) {
            this.text = text;
            directory.setText(text);
        }
    }
}

File: app/src/main/java/com/gnardini/tvshowcontroller/directories/DirectoryAdapter.kt
package com.gnardini.tvshowcontroller.directories

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.gnardini.tvshowcontroller.R

class DirectoryAdapter(private val listener: DirectoryClickListener)
    : RecyclerView.Adapter<DirectoryAdapter.DirectoryViewHolder>() {

    private var directories = listOf<String>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DirectoryViewHolder {
        val view = LayoutInflater
                .from(parent.context)
                .inflate(R.layout.directory_view, parent, false)
        return DirectoryViewHolder(view, listener)
    }

    override fun onBindViewHolder(holder: DirectoryViewHolder, position: Int) {
        holder.text = directories[position]
    }

    override fun getItemCount(): Int = directories.size

    fun setDirectories(directories: List<String>) {
        this.directories = directories
        notifyDataSetChanged()
    }

    class DirectoryViewHolder(
            itemView: View,
            private val listener: DirectoryClickListener)
        : RecyclerView.ViewHolder(itemView) {

        var directory: TextView = itemView.findViewById(R.id.directory)
        var text: String = ""
            set(value) {
                field = value
                directory.text = value
            }

        init {
            itemView.setOnClickListener { listener.invoke(text) }
        }
    }
}
